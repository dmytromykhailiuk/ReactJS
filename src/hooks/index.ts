export * from "./useInputValue";
export * from "./useLocalTime";
