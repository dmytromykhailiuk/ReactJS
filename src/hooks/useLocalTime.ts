import { useState, useEffect } from "react";

const getLocalTimeString = () => new Date().toLocaleTimeString("it-IT");

export function useLocalTime(delay = 1000) {
  const [localTime, changeLocalTime] = useState(getLocalTimeString());

  useEffect(() => {
    const interval = setInterval(() => {
      changeLocalTime(getLocalTimeString());
    }, delay);

    return () => {
      clearInterval(interval);
    };
  }, []);

  return localTime;
}
