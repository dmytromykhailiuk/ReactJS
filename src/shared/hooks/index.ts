export * from "./useInputValue";
export * from "./useClickOutside";
