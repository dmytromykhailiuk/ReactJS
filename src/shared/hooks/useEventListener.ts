import { EventTypes } from "../enums/event-types";
import { useEffect } from "react";

export function useEventListener(
  event: EventTypes,
  fn: (e: any) => void,
  deps: any[],
  shouldUse: boolean = true
): void {
  useEffect(() => {
    if (shouldUse) {
      document.addEventListener(event, fn);
      return () => {
        document.removeEventListener(event, fn);
      };
    }
  }, deps);
}
