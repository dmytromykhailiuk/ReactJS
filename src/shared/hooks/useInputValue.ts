import { useState, useCallback } from "react";

export function useInputValue(message: string) {
  const [value, changeValue] = useState(message);

  const onChangeValue = useCallback((event) => {
    changeValue(event.target.value);
  }, []);

  return {
    value,
    onChangeValue,
  };
}
