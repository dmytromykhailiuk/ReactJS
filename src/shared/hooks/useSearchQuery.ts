import { useLocation } from "react-router";

export function useSearchQuery() {
  return new URLSearchParams(useLocation().search).get("SearchQuery");
}
