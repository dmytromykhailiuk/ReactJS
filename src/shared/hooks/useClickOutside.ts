import { useEffect, useMemo } from "react";
import { clickOutside } from "shared/helpers";

export function useClickOutside(
  className: string,
  callback: () => void,
  shouldUse: boolean = true
) {
  const clickOutsides = useMemo(() => clickOutside(className, callback), []);

  useEffect(() => {
    if (shouldUse) {
      document.addEventListener("click", clickOutsides);
      return () => {
        document.removeEventListener("click", clickOutsides);
      };
    }
  }, [shouldUse]);
}
