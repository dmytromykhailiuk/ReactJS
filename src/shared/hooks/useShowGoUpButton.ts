import { useCallback, useState } from "react";
import { useEventListener } from "./useEventListener";
import { EventTypes } from "../enums";
import { getPageYOffsetValue } from "../helpers";

export function useShowGoUpButton(): boolean {
  const [showGoUpButton, setShowGoUpButtonValue] = useState<boolean>(false);

  const setShowGoUpButton = useCallback(() => {
    const pageYOffset = getPageYOffsetValue();
    if (!showGoUpButton && pageYOffset > 200) {
      setShowGoUpButtonValue(true);
    } else if (showGoUpButton && pageYOffset < 100) {
      setShowGoUpButtonValue(false);
    }
  }, [showGoUpButton]);

  useEventListener(EventTypes.SCROLL, setShowGoUpButton, [showGoUpButton]);

  return showGoUpButton;
}
