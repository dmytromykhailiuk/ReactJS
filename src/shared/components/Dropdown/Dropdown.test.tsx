import { shallow, mount } from 'enzyme';
import React from 'react';
import toJson from 'enzyme-to-json';
import { useField } from 'formik';
import { Dropdown } from '.';
import { useClickOutside } from '../../hooks';

jest.mock('formik');
jest.mock('../../hooks');

const setValue = jest.fn();
const setTouched = jest.fn();

describe('Dropdown', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    (useField as jest.Mock).mockImplementation(() => [null, { value: [] }, { setValue, setTouched }]);
    (useClickOutside as jest.Mock).mockImplementation(() => {});
  });
  it('should match snapshot', () => {
    const wrapper = shallow(
      <Dropdown name="dropdown" options={[]} label="label" placeholder="placeholder" error="error" />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call setTouched when closing', () => {
    (useField as jest.Mock).mockImplementation(() => [null, { value: ['s', 'q'] }, { setValue, setTouched }]);
    const wrapper = mount(<Dropdown name="dropdown" options={['q'] as any} />);
    wrapper.find('.interface').simulate('click').simulate('click');
    expect(setTouched).toHaveBeenCalled();
  });

  it('should call setValue', () => {
    (useField as jest.Mock).mockImplementation(() => [null, { value: ['s'] }, { setValue, setTouched }]);
    const wrapper = mount(<Dropdown name="dropdown" options={['q', 's'] as any} />);
    wrapper.find('.interface').simulate('click');
    wrapper.find('span').first().simulate('click');
    wrapper.find('span').last().simulate('click');
    expect(setValue).toHaveBeenCalledTimes(2);
  });

  it('should call setTouched when click ouside', () => {
    let shouldUpdate = true;
    (useClickOutside as jest.Mock).mockImplementation((_, fn) => {
      if (shouldUpdate) {
        fn();
        shouldUpdate = false;
      }
    });
    mount(<Dropdown name="dropdown" options={['q'] as any} />);
    expect(setTouched).toHaveBeenCalled();
  });
});
