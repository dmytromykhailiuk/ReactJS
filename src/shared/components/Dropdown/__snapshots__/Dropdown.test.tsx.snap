// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should match snapshot 1`] = `
<div
  className="dropdown"
>
  <div
    className="interface"
    onClick={[Function]}
  >
    <div
      className="dropdown__label"
    >
      label
    </div>
    <div
      className="dropdown__field dropdown__field--placeholder dropdown__field--error"
    >
      placeholder
      <div
        className="dropdown__arrow-down"
      />
    </div>
  </div>
  <div
    className="dropdown__error"
  >
    error
  </div>
  <div
    className="dropdown__options-wrapper"
  />
</div>
`;
